removed up & down arrows from cart

sojeri opened the pull request:
- removed up arrow from display when product has no available stock
- removed down arrow from display when order item's quantity is 1
- updated relevant order item specs

sojeri committed:
removed down & up arrows from cart. found & fixed down arrow bug when max qty in cart already. n_n

sojeri committed:
uhhhhh working except where it isn't. see line 193 in order_item_spec. more somehow increasing beyond available stock. wat do?

sojeri committed:
fixed bug in #more! by forcing ActiveRecord to reload the record.

catchingash commented on app/controllers/order_items_controller.rb:
```diff
@@ -4,11 +4,11 @@ class OrderItemsController < ApplicationController
       # !W we should talk about whether this error message should be shown
       # and what it should truly say in production
-      flash[:error] = "You cannot increase the quantity of #{ @item.product.name } any further, because there are only #{ @item.quantity_ordered } in stock."
+      flash[:error] = @item.errors[:quantity_ordered]
```
This would make more sense to me if it were just `@item.errors` in order to accommodate other errors that may happen.

sojeri replied:
Other errors should be adding themselves to the errors hash already, no? Or maybe I don't understand what you mean.

catchingash replied:
They're not adding themselves to the flash[:error], no.

catchingash commented on app/models/order_item.rb:
```diff
@@ -13,12 +13,26 @@ class OrderItem < ActiveRecord::Base
+    reload
+
+    # if the OrderItem isn't reloaded, this will resolve based on a cached operation
+    # by reloading, we force a new SQL query to check whether product_has_stock?
```
I'm not sure what this comment means.

catchingash commented on app/models/order_item.rb:
```diff
@@ -13,12 +13,26 @@ class OrderItem < ActiveRecord::Base
   end
 
+  def more!
+    reload
```
Why is this necessary?

sojeri replied:
Do you understand the comments now?

catchingash commented on app/models/order_item.rb:
```diff
@@ -13,12 +13,26 @@ class OrderItem < ActiveRecord::Base
+    update_column(:quantity_ordered, quantity_ordered - 1) if quantity_ordered > 1
+  end
+
   # group: note that we need to talk about this more
```
While you're in here, do you want to add a `TODO` in this comment?

catchingash commented on app/models/order_item.rb:
```diff
@@ -36,12 +50,13 @@ def item_price # v. price / cost / item_cost / total_item_price / total_item_cos
   def product_has_stock? # !Q is this the right way to do this?
     # stock = product.has_available_stock?
-    unless product.has_available_stock?
+    if product.has_available_stock?
```
Does this method return true/false? I think this could be refactored so that the result of .has_available_stock? is returned, and the error message set elsewhere?

sojeri replied:
How should the error message be set elsewhere? Like-- do you have an idea where that code should live?

catchingash replied:
I like the #optimize tag you added. That seems like the best approach here. I was envisioning some sort of validation or something, rather than a before_save or whatever. But this is fine for now.

sojeri committed:
adding more explicit comments

catchingash commented on app/views/orders/_display_mutable_order_items.html.erb:
```diff
@@ -8,11 +8,14 @@
-        <%= link_to "-", less_item_path(item), method: :patch %>
+      <td>
+        <% if item.quantity_ordered > 1 %>
+          <%= link_to "-", less_item_path(item), method: :patch %>
```
It's fine if you prefer links, but I like to follow the rule that a link should only be a GET request, and a button should be any other kind of request.

sojeri committed:
missed a couple TODO flags

catchingash commented on spec/models/order_item_spec.rb:
```diff
@@ -177,6 +177,34 @@
+        expect(item.quantity_ordered).to eq(1)
+      end
+    end
+
```
Is there a test to confirm that the quantity cannot be increased beyond available stock, or decreased below 1?

sojeri replied:
yes, lines 201-205 & lines 187-191

catchingash replied:
Perfect. Thanks! Clearly I'm being lazy today.

catchingash commented on app/models/order_item.rb:
```diff
@@ -13,13 +13,30 @@ class OrderItem < ActiveRecord::Base
-  # group: note that we need to talk about this more
+  def more!
+    reload # removing this line == DANGER WILL ROBINSON
+    # if the OrderItem isn't reloaded, ln28 will resolve based on a cached operation
```
do you mean ln 30? It might be better to just reference the method by name.

sojeri replied:
whoops. it was line 28 before I expanded the comment. >_>

catchingash commented on app/models/order_item.rb:
```diff
@@ -13,13 +13,30 @@ class OrderItem < ActiveRecord::Base
+    reload # removing this line == DANGER WILL ROBINSON
+    # if the OrderItem isn't reloaded, ln28 will resolve based on a cached operation
+    # in other words, if line 28 has recently been evaluated for this OrderItem,
+    # then __in will not be evaluated__ and the cached value (true) will be used
```
typo? 'in will not be evaluated'?

sojeri committed:
refactored code & updated specs

catchingash commented on spec/controllers/order_items_controller_spec.rb:
```diff
@@ -81,6 +81,7 @@
         end
 
         expect(assigns(:item).quantity_ordered).to eq(@max_quantity - already_tied_quantity)
+        expect(assigns(:item).errors.keys).to include(:quantity_ordered)
```
We still need a test for flash[:errors]. 

Also, making sure the errors includes quantity_ordered seems like a model test, not a controller test.

catchingash committed:
Removes unnecessary expectations, adds pending.

The removed expectations belong in the model, not controller.

Adds pending for missing flash[:errors]
